pymc/logprob/order.py

```diff
@@ -0,0 +1,114 @@
+#   Copyright 2023 The PyMC Developers
+#
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+#
+#   MIT License
+#
+#   Copyright (c) 2021-2022 aesara-devs
+#
+#   Permission is hereby granted, free of charge, to any person obtaining a copy
+#   of this software and associated documentation files (the "Software"), to deal
+#   in the Software without restriction, including without limitation the rights
+#   to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+#   copies of the Software, and to permit persons to whom the Software is
+#   furnished to do so, subject to the following conditions:
+#
+#   The above copyright notice and this permission notice shall be included in all
+#   copies or substantial portions of the Software.
+#
+#   THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+#   IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+#   FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+#   AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+#   LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+#   OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+#   SOFTWARE.
+
+from typing import List, Optional
+
+from pytensor.graph.basic import Node
+from pytensor.graph.fg import FunctionGraph
+from pytensor.graph.rewriting.basic import node_rewriter
+from pytensor.tensor.math import MaxAndArgmax
+
+from pymc.logprob.abstract import MeasurableVariable, _logcdf, _logprob
+from pymc.logprob.rewriting import measurable_ir_rewrites_db
+
+
+class MeasurableMax(MaxAndArgmax):
```

Comment:
Actually there seems to be a pure `Max` argument in PyTensor. So to make our lives easier, we can have a rewrite that replaces `MaxAndArgmax` by an equivalent `Max`, and then have this rewrite only have to worry about `Max`ed variables.

The Op is defined here: https://github.com/pymc-devs/pytensor/blob/ec6a3153fbf5aaa023368f36aac58212c9bedbd4/pytensor/tensor/math.py#L622-L630

There is also one for `Argmax`: https://github.com/pymc-devs/pytensor/blob/ec6a3153fbf5aaa023368f36aac58212c9bedbd4/pytensor/tensor/math.py#L331

This kind of intermediate rewrites that don't return a MeasurableVariable look like this: https://github.com/pymc-devs/pymc/blob/765fdef219ea821dcc8ddbbd4291dc42e739341b/pymc/logprob/transforms.py#L490-L494

Reply:
Actually the rewrite already exists here: https://github.com/pymc-devs/pytensor/blob/ec6a3153fbf5aaa023368f36aac58212c9bedbd4/pytensor/tensor/rewriting/uncanonicalize.py#L45C1-L60

We can simply register it in our pre-canonicalization phase

Reply:
> Do you mean to write a rewrite that converts MaxAndArgmax op to Max so that find_measurable_max works with Max instead of MaxAndArgmax? 

Yes

> If so what should be done when someone uses pt.argmax? Both pt.max and pt.argmax create a MaxAndArgmax op and if we do this, we will also convert MaxAndArgmax op to Max`` when pt.argmax``` is called

No, the rewrite that converts `Argmax` to `Max` only does so when only the firt output `max` is being used and not the second `argmax`. If a user asks for a `pt.argmax` that produced `MaxAndArgmax` will never be rewritten to a `Max`, as that would be wrong.

Reply:
Makes sense. Thanks for the clarification 😃